File: lib/helldivers_2/models/assignments/assignment/setting.ex
defmodule Helldivers2.Models.Assignments.Assignment.Setting do
  alias Helldivers2.Models.Assignments.Assignment.Reward
  alias Helldivers2.Models.Assignments.Assignment.Task

  @type t :: %__MODULE__{
    type: non_neg_integer(),
    overrideTitle: String.t(),
    overrideBrief: String.t(),
    taskDescription: String.t(),
    tasks: list(Task.t()),
    reward: Reward.t(),
    flags: non_neg_integer(),
  }

  defstruct [
    :type,
    :overrideTitle,
    :overrideBrief,
    :taskDescription,
    :tasks,
    :reward,
    :flags,
  ]

  def parse(map) when is_map(map) do
    %__MODULE__{
      type: Map.get(map, "type"),
      overrideTitle: Map.get(map, "overrideTitle"),
      overrideBrief: Map.get(map, "overrideBrief"),
      taskDescription: Map.get(map, "taskDescription"),
      tasks: Map.get(map, "tasks"),
      reward: Map.get(map, "reward"),
      flags: Map.get(map, "flags"),
    }
  end
end

File: src/Helldivers-2-Models/ArrowHead/Assignments/Setting.cs
namespace Helldivers.Models.ArrowHead.Assignments;

/// <summary>
/// Contains the details of an <see cref="Assignment" /> like reward and requirements.
/// </summary>
/// <param name="Type">The type of assignment, values unknown at the moment.</param>
/// <param name="OverrideTitle">The title of this assignment.</param>
/// <param name="OverrideBrief">The briefing (description) of this assignment.</param>
/// <param name="TaskDescription">A description of what is expected of Helldivers to complete the assignment.</param>
/// <param name="Tasks">A list of <see cref="Task" />s describing the assignment requirements.</param>
/// <param name="Reward">Contains information on the rewards players willr eceive upon completion.</param>
/// <param name="Flags">Flags, suspected to be a binary OR'd value, purpose unknown.</param>
public sealed record Setting(
    int Type,
    string OverrideTitle,
    string OverrideBrief,
    string TaskDescription,
    List<Task> Tasks,
    Reward Reward,
    int Flags
);
